### PR TITLE
Add module to gather facts of a SolidFire cluster.

### PR DESCRIPTION
**Pull request by hulquest**
##### SUMMARY

Add module to gather facts of a SolidFire cluster.  The module gathers:
- Cluster information
- List cluster nodes.
- Cluster state
- Accounts
- Drives
- Volumes
- Virtual Volumes (vvols)
- Snapshots
- Cluster Faults

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

ansible/modules/storage/netapp/sf_facts.py
##### ANSIBLE VERSION

```
✔ ~/src/projects/ansible [sf-facts|✚ 1…2] 
14:44 $ ansible --version
ansible 2.4.0 (sf-facts de9f032a10) last updated 2017/09/01 14:55:21 (GMT -600)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/khulques/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/khulques/src/projects/ansible/lib/ansible
  executable location = /home/khulques/src/projects/ansible/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]

```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: lib/ansible/modules/storage/netapp/sf_facts.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
+#!/usr/bin/python
+
+# (c) 2017, NetApp, Inc
+#
+# This file is part of Ansible
```
Comment (Akasurde) [inline lib/ansible/modules/storage/netapp/sf_facts.py]: Can you please replace line 4 to 19 to match a single liner GPL license (which is new in Ansible Module development)? 


```
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

File: lib/ansible/modules/storage/netapp/sf_facts.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+RETURN = """
+msg:
+    description: Gathered facts for Solidfire cluster.
+    returned: always
+    type: string
+"""
+
+import ansible.module_utils.netapp as netapp_utils
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.pycompat24 import get_exception
```
Comment (Akasurde) [inline lib/ansible/modules/storage/netapp/sf_facts.py]: get_exception is deprecated in favor of Python native exception handling. 

```
except Exception as e:
    module.fail_json(msg=to_native(e)
```

**Review by Akasurde — CHANGES_REQUESTED** [review]
Rest Looks good to me.

File: lib/ansible/modules/storage/netapp/sf_facts.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+                msg="Failed to obtain cluster fact [%s] from [%s]. Error [%s]" % (name, self.module.params['hostname'],
+                                                                                  to_native(sf_problem)))
+        except ApiServerError as problem:
+            self.module.fail_json(msg=to_native(problem))
+
+    def gather_facts(self):
+        """Gather information from the cluster and return it to Ansible."""
+
+        self.facts['mvip'] = self.module.params['hostname']
+        """This is the cluster identifyer."""
```
Comment (lmprice) [inline lib/ansible/modules/storage/netapp/sf_facts.py]: Typo here.

**Review by lmprice — APPROVED** [review]
Very minor nit, otherwise, LGTM.
shipit

File: lib/ansible/modules/storage/netapp/sf_facts.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                    'supported_by': 'community',
+                    'metadata_version': '1.1'}
+
+DOCUMENTATION = '''
+module: sf_facts
+
+short_description: Report facts of a NetApp SolidFire cluster.
+extends_documentation_fragment:
+    - netapp.solidfire
+version_added: '2.4'
```
Comment (Akasurde) [inline lib/ansible/modules/storage/netapp/sf_facts.py]: 2.4 branch is freezed for development so I guess version_added will be 2.5

Take a look at - [shippable](https://app.shippable.com/github/ansible/ansible/runs/36453/1/console)

**Comment by broncofan** [comment]
LGTM, 
shipit

**Review by Akasurde — APPROVED** [review]
shipit

File: lib/ansible/modules/storage/netapp/sf_facts.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+---
+    - name: Get cluster facts.
+      sf_facts:
+        hostname: "{{ solidfire_hostname }}"
+        username: "{{ solidfire_username }}"
+        password: "{{ solidfire_password }}"
+"""
+
+RETURN = """
+msg:
```
Comment (resmo) [inline lib/ansible/modules/storage/netapp/sf_facts.py:37]: msg is a common return and must not be documented. But your returns should be identical to the dict of facts returned.

**Review by resmo — CHANGES_REQUESTED** [review]

**Comment by hulquest** [comment]
Closing this PR since I messed up the rebase.  I'll reopen a new one.

**Pull request closed without merging**